MVVMIOSTests/NewsViewModelTests.swift

```diff
@@ -0,0 +1,59 @@
+//
+//  NewsViewModelTests.swift
+//  MVVMIOSTests
+//
+//  Created by Ravi Bastola on 2/18/22.
+//
+
+import XCTest
+@testable import MVVMIOS
+
+class NewsViewModelTests: XCTestCase {
+
+    var sut: NewsViewModel!
+
+    func testItGetsInitilizedWithAService() {
+        let mockService: MockNetworkClient = MockNetworkClient()
+        sut = NewsViewModel(service: mockService)
+        XCTAssertNotNil(sut.networkingService)
+    }
+
+    func testItThrowsInvalidUrlErrorWhenTheUrlIsInvalid() {
+        let mockService: MockNetworkClient = MockNetworkClient()
+        mockService.result = Result<News<[NewsData]>, Error>.failure(NetworkError.invalidURL as Error)
+        sut = NewsViewModel(service: mockService)
+        sut.fetchData(forURL: "http://invalid.url") { result in
+            switch result {
+            case .success:
+                XCTFail("No error thrown")
+            case .failure(let error):
+                XCTAssertEqual(NetworkError.invalidURL.localizedDescription, error.localizedDescription)
+            }
+        }
```

Comment:
If the object you are testing calls a completion, like in this case, it is best to store the result of that completion in an array and then assert on the value outside of the completion. The reason is that this test would pass even if you completely forget to call the completion inside your `fetchData` method. The way I usually do this is something like the below. This way you can assert that your closure was called exactly once, and that it was called with the failure (or success in future tests) that you expect. 
```suggestion
        var fetchDataResults: [Result<News<[NewsData]>, Error>] = []
        sut.fetchData(forURL: "http://invalid.url") { fetchDataResults.append($0) }
        XCTAssertEqual(fetchDataResults.count, 1)
        XCTAssertEqual(fetchDataResults[0], .failure(NetworkError.invalidURL))
```

Reply:
@tinder-owenthomas As, we are only pretending it to be synchronous, but our completion handler will work asynchronously In real life. is not the best candidate of expectations and fulfill scenario? I would want a more insight on using expectations and fulfillments.

Reply:
On the other note ` XCTAssertEqual(fetchDataResults[0], .failure(NetworkError.invalidURL))` This fails to compile.

Reply:
instead what I am thinking is 

```swift

struct ApplicationError: Equatable, Error {
    let message: String
}

```
Is this a good idea?

Reply:
Still no luck

Reply:
Yeah, my code as written may not compile due to needing more Type info, but the general idea works fine. As for expectations, that is one way to go but we generally discourage it internally since it allows for the possibility that tests are running asynchronously which is a huge waste of time. Also we've found that using lots of expectations can cause problems with CI and CD systems for some reason, so we have a rule against using them at all at work. However, both are generally valid approaches. Might not hurt to try coding up both options to make sure you're familiar with them and understand the differences, pros/cons, etc.